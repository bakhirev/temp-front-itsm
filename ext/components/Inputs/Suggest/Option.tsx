import React, {
  useState,
  useRef,
  useEffect,
  MutableRefObject,
  Dispatch,
  FC,
  MouseEvent,
  RefObject,
} from 'react';

import { createTooltipHOC } from '../../TooltipHOC';
import type { Size } from '../common';

import { isString } from './utils';
import {
  OptionWrapper,
  TextWrapper,
  StyledBody1Long,
  StyledBody2Long,
  ColorText,
} from './styled-components';

import type { IOptionSuggest } from './index';

interface IOption {
  size: Size;
  item?: string | IOptionSuggest<any>;
  tooltipMessage?: string;
  setShowMenu?: Dispatch<React.SetStateAction<boolean>>;
  tooltipContainer?: Element | null;
  inputRef?: RefObject<HTMLInputElement>;
  renderElementOption?: (options: string | IOptionSuggest<any>) => any;
  onChange?: (value: string | IOptionSuggest) => void;
  setInputValue?: Dispatch<React.SetStateAction<string>>;
  inputValue?: string;
  label?: string;
  selectedItem?: string | IOptionSuggest<any>;
}

export const Option: FC<IOption> = ({
  children,
  size,
  setShowMenu,
  tooltipContainer,
  inputRef,
  renderElementOption,
  item,
  setInputValue,
  onChange,
  inputValue,
  tooltipMessage,
  label,
  selectedItem,
}) => {
  const [overflow, setOverflow] = useState(false);

  const wrapperRef: MutableRefObject<HTMLDivElement | null> = useRef(null);
  const contentRef: MutableRefObject<HTMLDivElement | null> = useRef(null);

  const detectOverflow = (e: HTMLDivElement) => e?.offsetWidth < e?.scrollWidth;

  const Label = size === 'micro' ? StyledBody2Long : StyledBody1Long;
  const startText = label?.slice?.(0, inputValue?.length);
  const endText = label?.slice?.(inputValue?.length);

  useEffect(() => {
    const contentCurrent = contentRef.current;

    if (contentCurrent && detectOverflow(contentCurrent) !== overflow) {
      setOverflow(detectOverflow(contentCurrent));
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const OptionContent = overflow ? createTooltipHOC(TextWrapper) : TextWrapper;
  const tooltip = tooltipMessage || label || '';

  const onOptionClick = () => {
    inputRef?.current?.focus?.();
    setShowMenu?.(false);
    label && item && onChange?.(item);
    label && setInputValue?.(label);
  };

  const handleMouseDown = (e: MouseEvent<HTMLDivElement>) => e.preventDefault();

  const selected =
    !isString(item) && !isString(selectedItem)
      ? selectedItem?.id === item?.id
      : item === selectedItem;

  return (
    <OptionWrapper
      size={size}
      ref={wrapperRef}
      onMouseDown={handleMouseDown}
      onClick={onOptionClick}
    >
      {renderElementOption && item ? (
        renderElementOption(item)
      ) : (
        <OptionContent tooltip={tooltip} target={wrapperRef.current} container={tooltipContainer}>
          <Label ref={contentRef}>
            {label ? (
              <>
                {selected ? (
                  <ColorText>{label}</ColorText>
                ) : (
                  <>
                    <ColorText>{startText}</ColorText>
                    {endText}
                  </>
                )}
              </>
            ) : (
              children
            )}
          </Label>
        </OptionContent>
      )}
    </OptionWrapper>
  );
};
