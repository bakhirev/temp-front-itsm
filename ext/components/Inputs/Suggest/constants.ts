export const CHARS_TO_SHOW_SUGGEST = 1;
export const MENU_WRAPPER_Z_INDEX = 5;
export const MENU_PADDING = 8;
export const MAX_OPTIONS_BEFORE_SCROLL = 3;
export const OPTION_HORIZONTAL_PADDING = 18;
export const DEFAULT_NOOP_MESSAGE = 'Варианты не найдены';
export const DEFAULT_LOAD_MESSAGE = 'Поиск совпадений';
export const DEFAULT_MENU_HEIGHT = 190;
export const OPTION_SIZE = {
  big: 56,
  small: 56,
  micro: 48,
};
