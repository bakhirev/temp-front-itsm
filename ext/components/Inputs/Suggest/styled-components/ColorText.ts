import styled from 'styled-components';

import { DEFAULT_THEME } from '../../../common';

export const ColorText = styled.span`
  color: ${({ theme }) => theme.color.primary[60]};
`;

ColorText.defaultProps = {
  theme: DEFAULT_THEME,
};
