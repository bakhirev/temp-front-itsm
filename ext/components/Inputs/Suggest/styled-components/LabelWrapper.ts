import styled from 'styled-components';

import { DEFAULT_THEME } from '../../../common';

export const LabelWrapper = styled.div<{ load?: boolean }>`
  color: ${({ theme }) => theme.color.neutral[50]};
  padding-left: ${({ load }) => (load ? 13 : 0)}px;
`;

LabelWrapper.defaultProps = {
  theme: DEFAULT_THEME,
};
