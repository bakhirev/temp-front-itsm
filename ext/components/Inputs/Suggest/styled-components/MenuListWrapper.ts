import styled from 'styled-components';

import { DEFAULT_THEME } from '../../../common';
export const MenuListWrapper = styled.div<{
  load?: boolean;
  withPadding?: boolean;
  menuMaxHeight: string | number;
}>`
  -webkit-overflow-scrolling: touch;
  box-sizing: border-box;
  position: relative;
  padding-top: ${({ load, withPadding }) => (load || withPadding ? 0 : 8)}px;
  padding-bottom: ${({ load, withPadding }) => (load || withPadding ? 0 : 8)}px;
  overflow: hidden;
  overflow-y: auto;
  max-height: ${({ menuMaxHeight }) =>
    typeof menuMaxHeight === 'string' ? menuMaxHeight : `${menuMaxHeight}px`};
  ::-webkit-scrollbar {
    width: 16px;
  }

  ::-webkit-scrollbar-track {
    background: ${({ theme }) => theme.color.neutral.white};
    background-clip: content-box;
  }

  ::-webkit-scrollbar-thumb {
    background: ${({ theme }) => theme.color.neutral[30]};
    border: 6px solid ${({ theme }) => theme.color.neutral.white};
    border-radius: 10px;
  }

  ::-webkit-scrollbar-thumb:hover {
    border: 5px solid ${({ theme }) => theme.color.neutral.white};
  }
`;

MenuListWrapper.defaultProps = {
  theme: DEFAULT_THEME,
};
