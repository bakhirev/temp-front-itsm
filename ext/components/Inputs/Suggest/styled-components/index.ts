export * from './MenuListWrapper';
export * from './MenuWrapper';
export * from './StyledBodyLong';
export * from './IconWrapper';
export * from './SpinnerWrapper';
export * from './LabelWrapper';
export * from './StyledSearchOutline';
export * from './Wrapper';
export * from './InputWrapper';
export * from './ColorText';
export * from './OptionWrapper';
export * from './TextWrapper';
