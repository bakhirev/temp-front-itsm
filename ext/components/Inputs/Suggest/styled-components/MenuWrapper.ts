import styled, { css } from 'styled-components';

import { MENU_WRAPPER_Z_INDEX } from '../constants';
import { DEFAULT_THEME } from '../../../common';

export const MenuWrapper = styled.div<{
  menuMaxHeight: string | number;
  visible?: boolean;
}>`
  display: ${({ visible }) => (visible ? 'block' : 'none')};
  position: absolute;
  top: 100%;
  width: 100%;
  max-height: ${({ menuMaxHeight }) =>
    typeof menuMaxHeight === 'string' ? menuMaxHeight : `${menuMaxHeight}px`};
  box-sizing: border-box;
  z-index: ${MENU_WRAPPER_Z_INDEX};
  ${({ theme }) => css`
    margin-top: ${theme.input.dropdownMenuPadding}px;
    border-radius: ${theme.borderRadius};
    background-color: ${theme.color.neutral.white};
    box-shadow: ${theme.boxShadow[12]};
  `}
`;

MenuWrapper.defaultProps = {
  theme: DEFAULT_THEME,
};
