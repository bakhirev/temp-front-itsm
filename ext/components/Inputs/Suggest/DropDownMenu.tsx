import React, { FC, Dispatch, RefObject } from 'react';

import { Spinner } from '../../Spinner';
import type { Size } from '../common';

import { isString } from './utils';
import { MenuListWrapper, MenuWrapper, SpinnerWrapper, LabelWrapper } from './styled-components';
import { Option } from './Option';

import type { IOptionSuggest } from './index';
interface IDropDownMenu {
  showMenu: boolean;
  load: boolean;
  options: (string | IOptionSuggest<any>)[];
  setShowMenu: Dispatch<React.SetStateAction<boolean>>;
  noopMessage: string;
  loadMessage: string;
  tooltipContainer?: Element | null;
  size: Size;
  inputRef?: RefObject<HTMLInputElement>;
  menuMaxHeight: string | number;
  menuIsOpen?: boolean;
  renderElementOption?: (options: string | IOptionSuggest<any>) => any;
  onChange: (value: string | IOptionSuggest<any>) => void;
  setInputValue: Dispatch<React.SetStateAction<string>>;
  inputValue: string;
  selectedItem?: string | IOptionSuggest<any>;
  shouldRenderDropdown: boolean;
}

export const DropDownMenu: FC<IDropDownMenu> = ({
  load,
  options,
  size,
  noopMessage,
  loadMessage,
  tooltipContainer,
  inputRef,
  menuMaxHeight,
  renderElementOption,
  onChange,
  menuIsOpen,
  setShowMenu,
  setInputValue,
  inputValue,
  selectedItem,
  shouldRenderDropdown,
}) => {
  const withPadding = options.length === 0 || options.length === 1;

  const shouldRenderNoopMessage = options.length === 0 && !load;
  const shouldRenderOptions = options.length > 0 && !load;
  const spinnerSize = size === 'micro' ? 'small' : 'medium';

  return (
    <MenuWrapper menuMaxHeight={menuMaxHeight} visible={shouldRenderDropdown}>
      <MenuListWrapper load={load} withPadding={withPadding} menuMaxHeight={menuMaxHeight}>
        {load && (
          <Option size={size} tooltipMessage={loadMessage}>
            <SpinnerWrapper>
              <Spinner size={spinnerSize} />
              <LabelWrapper load>{loadMessage}</LabelWrapper>
            </SpinnerWrapper>
          </Option>
        )}
        {shouldRenderOptions &&
          options.map((item: string | IOptionSuggest<any>, index) => (
            <Option
              key={isString(item) ? index : item.id}
              size={size}
              tooltipContainer={tooltipContainer}
              inputRef={inputRef}
              setShowMenu={menuIsOpen === undefined ? setShowMenu : undefined}
              renderElementOption={renderElementOption}
              onChange={onChange}
              label={isString(item) ? item : item.label}
              item={item}
              setInputValue={setInputValue}
              inputValue={inputValue}
              selectedItem={selectedItem}
            />
          ))}
        {shouldRenderNoopMessage && (
          <Option size={size} tooltipMessage={noopMessage}>
            <LabelWrapper>{noopMessage}</LabelWrapper>
          </Option>
        )}
      </MenuListWrapper>
    </MenuWrapper>
  );
};
