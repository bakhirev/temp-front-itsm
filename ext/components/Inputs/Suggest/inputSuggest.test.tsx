import React from 'react';
import { mount, shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';

import { Input } from '../index';
import { InputSkeleton } from '../BaseField/InputSkeleton';
import { InputComponent } from '../BaseField/StyledComponents';

import { DEFAULT_NOOP_MESSAGE, DEFAULT_LOAD_MESSAGE } from './constants';
import { OptionWrapper, TextWrapper } from './styled-components';
import { DropDownMenu } from './DropDownMenu';
import { Option } from './Option';

import type { IOptionSuggest } from './index';

describe('Input.Suggest', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const inputSuggestProps = {
    options: ['testValue', 'testValue'],
    onChange: (value: any) => value,
  };
  const testValue = [
    { id: 1, label: 'Москва' },
    { id: 2, label: 'Мо' },
    { id: 3, label: 'Мос' },
    { id: 4, label: 'LA' },
    { id: 6, label: 'Chicago' },
  ];

  it('should render input suggest', () => {
    const wrapper = shallow(<Input.Suggest {...inputSuggestProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render input suggest with small size', () => {
    const wrapper = shallow(<Input.Suggest {...inputSuggestProps} size={'small'} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render input suggest with micro size', () => {
    const wrapper = shallow(<Input.Suggest {...inputSuggestProps} size={'micro'} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onChange when user presses keys in input field', () => {
    const changeSpy = jest.fn();
    const wrapper = mount(<Input.Suggest {...inputSuggestProps} onInputChange={changeSpy} />);
    const input = wrapper.find(InputSkeleton).find(InputComponent);
    input.simulate('change', { target: { value: 'Changed' } });
    expect(changeSpy).toHaveBeenCalled();
    expect(changeSpy).toBeCalledTimes(1);
  });

  it('should render input suggest with Array object', () => {
    const wrapper = shallow(
      <Input.Suggest
        options={testValue}
        onChange={(value) => value}
        menuIsOpen
        charsToShowSuggest={0}
      />
    );
    const option = (index) =>
      wrapper.find(DropDownMenu).dive().children().childAt(index).prop('label');
    expect(option(0)).toBe('Москва');
    expect(option(1)).toBe('Мо');
    expect(option(2)).toBe('Мос');
    expect(option(3)).toBe('LA');
    expect(option(4)).toBe('Chicago');
  });

  it('should render input suggest with custom option', () => {
    const wrapper = shallow(
      <Input.Suggest
        options={testValue}
        renderElementOption={(options: any) => (
          <div>
            <button>Нажми меня</button>
            {options.id} {options.label}
          </div>
        )}
        onChange={(value) => value}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render input suggest one option', () => {
    const changeSpy = jest.fn();
    const wrapper = mount(
      <Input.Suggest onInputChange={changeSpy} options={testValue} onChange={(value) => value} />
    );

    const input = wrapper.find(InputSkeleton).find(InputComponent);
    input.simulate('change', { target: { value: 'Москва' } });
    expect(changeSpy).toHaveBeenCalled();
    expect(changeSpy).toBeCalledTimes(1);
    const option = wrapper.find(DropDownMenu).find(Option);
    expect(option.prop('label')).toBe('Москва');
    expect(option.children()).toHaveLength(1);
  });

  it('It should render input suggest options if there is match', () => {
    const changeSpy = jest.fn();
    const wrapper = mount(
      <Input.Suggest
        onInputChange={changeSpy}
        menuIsOpen
        charsToShowSuggest={0}
        options={testValue}
        onChange={(value) => value}
      />
    );

    const input = wrapper.find(InputSkeleton).find(InputComponent);
    input.simulate('change', { target: { value: 'Мо' } });
    expect(changeSpy).toHaveBeenCalled();
    expect(changeSpy).toBeCalledTimes(1);
    const option = wrapper.find(DropDownMenu).find(Option);
    expect(option.children()).toHaveLength(3);
  });

  it('should render input suggest noop message,', () => {
    const changeSpy = jest.fn();
    const wrapper = mount(
      <Input.Suggest
        onInputChange={changeSpy}
        menuIsOpen
        charsToShowSuggest={0}
        options={testValue}
        onChange={(value) => value}
      />
    );

    const input = wrapper.find(InputSkeleton).find(InputComponent);
    input.simulate('change', { target: { value: 'Test' } });
    expect(changeSpy).toHaveBeenCalled();
    expect(changeSpy).toBeCalledTimes(1);
    const option = wrapper.find(DropDownMenu).find(Option);
    expect(option.children()).toHaveLength(1);
    expect(option.prop('tooltipMessage')).toBe(DEFAULT_NOOP_MESSAGE);
  });

  it('should render input suggest load message', () => {
    const changeSpy = jest.fn();
    const wrapper = mount(
      <Input.Suggest
        onInputChange={changeSpy}
        menuIsOpen
        charsToShowSuggest={0}
        options={testValue}
        load
        onChange={(value) => value}
      />
    );

    const input = wrapper.find(InputSkeleton).find(InputComponent);
    input.simulate('change', { target: { value: 'Test' } });
    expect(changeSpy).toHaveBeenCalled();
    expect(changeSpy).toBeCalledTimes(1);
    const option = wrapper.find(DropDownMenu).find(Option);
    expect(option.children()).toHaveLength(1);
    expect(option.prop('tooltipMessage')).toBe(DEFAULT_LOAD_MESSAGE);
  });

  it('should call onChange when user click option', () => {
    const changeSpy = jest.fn();
    let resultValue: string | IOptionSuggest = '';
    const wrapper = mount(
      <Input.Suggest
        onInputChange={changeSpy}
        options={testValue}
        menuIsOpen
        charsToShowSuggest={0}
        onChange={(value: string | IOptionSuggest) => (resultValue = value)}
      />
    );

    const input = wrapper.find(InputSkeleton).find(InputComponent);
    input.simulate('change', { target: { value: 'Мо' } });
    expect(changeSpy).toHaveBeenCalled();
    expect(changeSpy).toBeCalledTimes(1);
    const options = wrapper.find(DropDownMenu).find(Option);
    options.forEach((option, index) => {
      if (index === 0) {
        act(() => {
          option.find(OptionWrapper).find(TextWrapper).simulate('click');
        });
      }
    });

    expect(JSON.stringify(resultValue)).toBe(JSON.stringify(testValue[0]));
  });
});
