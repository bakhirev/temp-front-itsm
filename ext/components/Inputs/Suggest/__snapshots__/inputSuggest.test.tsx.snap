// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input.Suggest should render input suggest 1`] = `
<styled.div>
  <Component
    inputRef={
      Object {
        "current": null,
      }
    }
    onChangeSelected={[Function]}
    setInputValue={[Function]}
    setShowMenu={[Function]}
    size="big"
    value=""
  />
  <Component
    inputRef={
      Object {
        "current": null,
      }
    }
    inputValue=""
    load={false}
    loadMessage="Поиск совпадений"
    menuMaxHeight={190}
    noopMessage="Варианты не найдены"
    onChange={[Function]}
    options={
      Array [
        "testValue",
        "testValue",
      ]
    }
    setInputValue={[Function]}
    setShowMenu={[Function]}
    shouldRenderDropdown={false}
    showMenu={false}
    size="big"
  />
</styled.div>
`;

exports[`Input.Suggest should render input suggest with custom option 1`] = `
<styled.div>
  <Component
    inputRef={
      Object {
        "current": null,
      }
    }
    onChangeSelected={[Function]}
    setInputValue={[Function]}
    setShowMenu={[Function]}
    size="big"
    value=""
  />
  <Component
    inputRef={
      Object {
        "current": null,
      }
    }
    inputValue=""
    load={false}
    loadMessage="Поиск совпадений"
    menuMaxHeight={190}
    noopMessage="Варианты не найдены"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 1,
          "label": "Москва",
        },
        Object {
          "id": 2,
          "label": "Мо",
        },
        Object {
          "id": 3,
          "label": "Мос",
        },
        Object {
          "id": 4,
          "label": "LA",
        },
        Object {
          "id": 6,
          "label": "Chicago",
        },
      ]
    }
    renderElementOption={[Function]}
    setInputValue={[Function]}
    setShowMenu={[Function]}
    shouldRenderDropdown={false}
    showMenu={false}
    size="big"
  />
</styled.div>
`;

exports[`Input.Suggest should render input suggest with micro size 1`] = `
<styled.div>
  <Component
    inputRef={
      Object {
        "current": null,
      }
    }
    onChangeSelected={[Function]}
    setInputValue={[Function]}
    setShowMenu={[Function]}
    size="micro"
    value=""
  />
  <Component
    inputRef={
      Object {
        "current": null,
      }
    }
    inputValue=""
    load={false}
    loadMessage="Поиск совпадений"
    menuMaxHeight={190}
    noopMessage="Варианты не найдены"
    onChange={[Function]}
    options={
      Array [
        "testValue",
        "testValue",
      ]
    }
    setInputValue={[Function]}
    setShowMenu={[Function]}
    shouldRenderDropdown={false}
    showMenu={false}
    size="micro"
  />
</styled.div>
`;

exports[`Input.Suggest should render input suggest with small size 1`] = `
<styled.div>
  <Component
    inputRef={
      Object {
        "current": null,
      }
    }
    onChangeSelected={[Function]}
    setInputValue={[Function]}
    setShowMenu={[Function]}
    size="small"
    value=""
  />
  <Component
    inputRef={
      Object {
        "current": null,
      }
    }
    inputValue=""
    load={false}
    loadMessage="Поиск совпадений"
    menuMaxHeight={190}
    noopMessage="Варианты не найдены"
    onChange={[Function]}
    options={
      Array [
        "testValue",
        "testValue",
      ]
    }
    setInputValue={[Function]}
    setShowMenu={[Function]}
    shouldRenderDropdown={false}
    showMenu={false}
    size="small"
  />
</styled.div>
`;
